yt_dlp/downloader/niconico.py

```diff
@@ -56,34 +67,33 @@ def heartbeat():
         return success
 
 
-class NiconicoLiveFD(FileDownloader):
-    """ Downloads niconico live without being stopped """
+class NiconicoLiveBaseFD(FileDownloader):
+    _WEBSOCKET_RECONNECT_DELAY = 10
 
-    def real_download(self, filename, info_dict):
-        video_id = info_dict['video_id']
-        ws_url = info_dict['url']
-        ws_extractor = info_dict['ws']
-        ws_origin_host = info_dict['origin']
-        live_quality = info_dict.get('live_quality', 'high')
-        live_latency = info_dict.get('live_latency', 'high')
-        dl = FFmpegFD(self.ydl, self.params or {})
-
-        new_info_dict = info_dict.copy()
-        new_info_dict.update({
-            'protocol': 'm3u8',
-        })
+    @contextlib.contextmanager
+    def _ws_context(self, info_dict):
+        """ Hold a WebSocket object and release it when leaving """
+
+        video_id = info_dict['id']
+        live_latency = info_dict['live_latency']
+        self.ws = info_dict['__ws']
+
+        self.m3u8_lock = threading.Event()
+        self.m3u8_url = info_dict['manifest_url']
+        self.m3u8_lock.set()
 
         def communicate_ws(reconnect):
             if reconnect:
-                ws = self.ydl.urlopen(Request(ws_url, headers={'Origin': f'https://{ws_origin_host}'}))
+                self.ws = self.ydl.urlopen(Request(
+                    self.ws.url, headers={'Origin': self.ws.wsw.request.headers['Origin']}))
```

Comment:
`wsw` is internal only  (part of the websockets library handler) and not part of the websocket response interface, so this will break when we introduce a new library.

(Sorry I've been needing to rename it)

Reply:
Added to that, if it is needed we can prob add the original Request object to WebSocket responses

Reply:
Hi, @coletdjnz ! Thanks for your comment.

As a high-level class, using internal things is not OK. In the original code, the hostname comes from IE. I can change it to that.

```Python
# Info Extractor
def _real_extract(self, url):
  return {
    "__ws": {
      "ws": ws,
      "origin": f'https://{hostname}',
    },
  }

# Downloader
self.ws.url, headers={'Origin': self.ws['origin']}))

```

Your opinion?

Reply:
Perhaps in `http_headers` infodict property?

Otherwise I'd say that is probably fine too, since it's internal

Reply:
> Perhaps in `http_headers` infodict property?

Done. Please see 41c6125907f59ee84b392d832299a6927b99a9c9 .